---
jupyter:
  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.14.5
  kernelspec:
    display_name: Python 3.7.12 64-bit (conda)
    language: python
    name: python3712jvsc74a57bd0dc0fe05456373cce17991fe9e2e9264df4f4d1e972d77814a9700f21c9e7a8e2
---

# Settings

```python
import fastq
import pysam
import gzip
from collections import Counter
import os
import pandas as pd
import numpy as np
from Bio.Seq import Seq
from Bio.pairwise2 import format_alignment
from pyfaidx import Fasta
from Bio import motifs
import logomaker
import matplotlib.pyplot as plt
import pyfaidx

```

## modules

```python
def read_vcf(df_path):
    vcf_cols = ['chrom1', 'pos1', 'ID', 'ref', 'alt', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'sample']
    df = pd.read_table(df_path, comment='#', names=vcf_cols, dtype={'chrom':str, 'pos1':float})
    
    return df
def str_to_fasta(seq):
    fastaseq = ''
    length = len(seq)
    for start in range(0, length, 80):
        end = min(start+80, length)
        subseq = seq[start:end]
        fastaseq += f'{subseq}\n'
    return fastaseq
def get_secondaries(read):
    secondaries = [a[1].split(';') for a in read.tags if a[0] == 'SA']
    if len(secondaries) > 0:
        secondaries = secondaries[0]
        secondaries = list(filter(lambda x: x!='', secondaries))
        return secondaries
    return []
def get_secondary_aligned_seq(secondary, ncseq):
    query_consumers = 'MIS=X' # https://samtools.github.io/hts-specs/SAMv1.pdf
    query_aligners = 'MI=X' # removing clip from query consumers
    clen_sum = 0
    secondary_aligned_seq = ''
    tname, tpos, tstrand, tcigar, _, _ = secondary.split(',')
    cigar_tuple = re.findall('(\d+)([A-Z])', tcigar)
    for ix, (clen, ctag) in enumerate(cigar_tuple):
        print(clen, ctag)
        clen = int(clen)
        if ctag in query_aligners:
            aligned_seq = ncseq[clen_sum:clen_sum+clen]
            secondary_aligned_seq += aligned_seq
        if ctag in query_consumers:
            clen_sum += clen
    return (tname, tpos, tstrand, tcigar, secondary_aligned_seq)

def read_fastq(path):
    """
    path: path to fastq[.gz]
    returns: dict[rname] = seq
    """
    rname_to_seq = {}
    rname_to_qual = {}
    fos = fastq.read(path)
    for fo in fos:
        rname = fo.head.split(' ')[0][1:]
        seq = fo.body
        qual = fo.qstr
        rname_to_seq[rname] = seq
        rname_to_qual[rname] = qual
    return rname_to_seq, rname_to_qual

def read_chromosome_lengths(genome_fasta_index):
    fai = pd.read_csv(genome_fasta_index, sep='\t', header=None, names=['chrom', 'length', 'V3', 'V4', 'V5'])
    fai = fai.set_index('chrom')['length']
    # print(fai.to_dict())
    return fai.to_dict()

class RefGenomeInfo(object):
    def __init__(self, version):
        if version == 'hg19':
            self.chromosomes = [str(a) for a in range(1, 23)] + ['X', 'Y']

            genome_fasta_index = pkg_resources.resource_filename('wgs_analysis', 'data/GRCh37-lite.fa.fai')

            self.chromosome_lengths = pd.Series(read_chromosome_lengths(genome_fasta_index)).reindex(self.chromosomes).astype(int)
            self.chromosome_lengths.index.name = 'chr'

            self.chromosome_end = np.cumsum(self.chromosome_lengths)
            self.chromosome_start = self.chromosome_end.shift(1)
            self.chromosome_start[0] = 0
            self.chromosome_start = self.chromosome_start.astype(int)
            self.chromosome_mid = (self.chromosome_start + self.chromosome_end) / 2.

            self.chromosome_info = pd.DataFrame({
                'chromosome_length': self.chromosome_lengths,
                'chromosome_end': self.chromosome_end,
                'chromosome_start': self.chromosome_start,
                'chromosome_mid': self.chromosome_mid,
            }).reset_index()

        elif version == 'hg38':
            self.chromosomes = [str(a) for a in range(1, 23)] + ['X', 'Y']
            self.chromosomes = ['chr'+c for c in self.chromosomes]

            # genome_fasta_index = pkg_resources.resource_filename('wgs_analysis', 'data/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai')
            genome_fasta_index = '/juno/work/shah/users/chois7/packages/wgs_analysis/wgs_analysis/data/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai'

            self.chromosome_lengths = pd.Series(read_chromosome_lengths(genome_fasta_index)).reindex(self.chromosomes).astype(int)
            self.chromosome_lengths.index.name = 'chr'
            # self.chromosome_lengths.index = self.chromosome_lengths.index.str.replace('chr', '')

            self.chromosome_end = np.cumsum(self.chromosome_lengths)
            self.chromosome_start = self.chromosome_end.shift(1)
            self.chromosome_start[0] = 0
            self.chromosome_start = self.chromosome_start.astype(int)
            self.chromosome_mid = (self.chromosome_start + self.chromosome_end) / 2.

            self.chromosome_info = pd.DataFrame({
                'chromosome_length': self.chromosome_lengths,
                'chromosome_end': self.chromosome_end,
                'chromosome_start': self.chromosome_start,
                'chromosome_mid': self.chromosome_mid,
            }).reset_index()
            
            
```

# Vector analyses

```python
chromosomes = ['chr'+str(i) for i in range(1, 22+1)] + ['chrX', 'chrY']
vectors = ['PBEF1NeoTransposon', 'GFPVector', 'puro-GFP-PGBD5_seq', 'PiggyBacVector']
chromosomes += vectors
len(chromosomes)
```

```python
bam_dir = '/juno/work/shah/users/chois7/tickets/sarcoma89/14472B/results/bam'
samples = ['14472B_201', '14472B_202', '14472B_500', '14472B_501']
bam_paths = {
    sample: f'{bam_dir}/{sample}.bam' for sample in samples
}
# vcf_data = {k: read_vcf(data_paths[k]) for k in data_paths}
```

## save rnames for read aligned to vectors

```python
for ix, sample in enumerate(samples):
    if '_20' in sample:
        refs = ['PBEF1NeoTransposon', 'PGBD5Vector']
    elif '_50' in sample:
        refs = ['DelPBEF1NeoTransposon', 'PGBD5Vector']
    else:
        raise ValueError(f'{ix}, {sample}')
        
    bam_path = bam_paths[sample]
    bam = pysam.AlignmentFile(bam_path, 'rb')
    rnames = set()
    for ref in refs:
        for read in bam.fetch(ref):
            rnames.add(read.qname)
    print(sample, 'done')
            
    with open(f'/juno/work/shah/users/chois7/tickets/sarcoma89/14472B/results/rnames/{sample}.rnames.txt', 'w') as out:
        content = '\n'.join(list(rnames)) + '\n'
        out.write(content)
```
